proxy/src/control/cache.rs

```diff
@@ -0,0 +1,199 @@
+pub struct Cache<T> {
+    values: HashSet<T>,
+    reset_on_next_modification: bool,
+}
+
+pub enum Exists<T> {
+    Unknown, // Unknown if the item exists or not
+    Yes(T),
+    No, // Affirmatively known to not exist.
+}
+
+pub enum CacheChange {
+    Insertion,
+    Removal,
+}
+
+// ===== impl Exists =====
+
+impl<T> Exists<T> {
+    pub fn take(&mut self) -> Exists<T> {
+        mem::replace(self, Exists::Unknown)
+    }
+}
+
+// ===== impl Cache =====
+
+impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
+    pub fn new() -> Self {
+        Cache {
+            values: HashSet::new(),
+            reset_on_next_modification: true,
+        }
+    }
+
+    pub fn values(&self) -> &HashSet<T> {
+        &self.values
+    }
+
+    pub fn set_reset_on_next_modification(&mut self) {
+        self.reset_on_next_modification = true;
+    }
+
+    pub fn extend<I, F>(&mut self, iter: I, on_change: &mut F)
+        where I: Iterator<Item = T>,
+              F: FnMut(T, CacheChange),
+    {
+        fn extend_inner<T, I, F>(values: &mut HashSet<T>, iter: I, on_change: &mut F)
+            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        {
+            for value in iter {
+                if values.insert(value) {
+                    on_change(value, CacheChange::Insertion);
+                }
+            }
+        }
+
+        if !self.reset_on_next_modification {
+            extend_inner(&mut self.values, iter, on_change);
+        } else {
+            let to_insert = iter.collect::<HashSet<T>>();
+            extend_inner(&mut self.values, to_insert.iter().map(|value| *value), on_change);
+            self.retain(&to_insert, on_change);
+        }
+        self.reset_on_next_modification = false;
+    }
+
+    pub fn remove<I, F>(&mut self, iter: I, on_change: &mut F)
+        where I: Iterator<Item = T>,
+              F: FnMut(T, CacheChange)
+    {
+        if !self.reset_on_next_modification {
+            for value in iter {
+                if self.values.remove(&value) {
+                    on_change(value, CacheChange::Removal);
+                }
+            }
+        } else {
+            self.clear(on_change);
+        }
+        self.reset_on_next_modification = false;
+    }
+
+    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
+        self.retain(&HashSet::new(), on_change)
+    }
+
+    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
+        where F: FnMut(T, CacheChange)
+    {
+        self.values.retain(|value| {
+            let retain = to_retain.contains(&value);
+            if !retain {
+                on_change(*value, CacheChange::Removal)
+            }
+            retain
+        });
+        self.reset_on_next_modification = false;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn cache_extend_reset_on_next_modification() {
```

Comment:
Very minor, but we may want to drop the `cache_` from the beginning of the names of these tests, since they're now namespaced in the `cache` module and will show up as `cache::cache_extend_reset_on_next_modification` and so on in test output.